### PR TITLE
[FXML-1991] Update Add and Sub Constant Folding

### PR DESCRIPTION
**Pull request by cmcgirr-amd**
Often times the statements between two casts which represent simple Quantize and Dequantize operations can be simplified as their zero points are zero.
```
    %0 = "tosa.const"() {value = dense<0> : tensor<1x1x1x1xi8>} : () -> tensor<1x1x1x1xi8>
    %1 = "tosa.cast"(%arg0) : (tensor<32x3x224x224xf32>) -> tensor<32x3x224x224xi8>
    %2 = "tosa.add"(%1, %0) {OutputName = "QuantizeLinear_2"} : (tensor<32x3x224x224xi8>, tensor<1x1x1x1xi8>) -> tensor<32x3x224x224xi8>
    %3 = "tosa.sub"(%2, %0) {LayerName = "DequantizeLinear_5"} : (tensor<32x3x224x224xi8>, tensor<1x1x1x1xi8>) -> tensor<32x3x224x224xi8>
    %4 = "tosa.cast"(%3) : (tensor<32x3x224x224xi8>) -> tensor<32x3x224x224xf32>
```
The add and sub statements are then redundant and do no work as the constants to these operations are zero. The add and sub operations can then be removed to simplify the IR for pattern matching.

### CODE REVIEW

File: mlir/test/Dialect/Tosa/constant-op-fold.mlir [hunk 1]
```diff
@@ -164,6 +164,16 @@ func.func @fold_add_splat_f32() -> tensor<10xf32> {
...
 // -----
 
+// CHECK-LABEL: @fold_add_zero_splat_different_shape_f32
+func.func @fold_add_zero_splat_different_shape_f32(%arg0: tensor<1x10xf32>) -> tensor<1x10xf32> {
+  %zero = "tosa.const"() {value = dense<0.0> : tensor<1x1xf32>} : () -> tensor<1x1xf32>
+  %sub = "tosa.add"(%arg0, %zero) : (tensor<1x10xf32>, tensor<1x1xf32>) -> tensor<1x10xf32>
+  // CHECK: return %arg0
+  return %sub : tensor<1x10xf32>
+}
+
```
Comment (mgehre-amd) [inline mlir/test/Dialect/Tosa/constant-op-fold.mlir:174]: Can you add a test case where the zero constant causes broadcasting to happen, i.e.
```
func.func @fold_add_zero_splat_different_shape_f32(%arg0: tensor<1x10xf32>) -> tensor<1x10xf32> {
  %zero = "tosa.const"() {value = dense<0.0> : tensor<1x1xf32>} : () -> tensor<4x10xf32>
  %sub = "tosa.add"(%arg0, %zero) : (tensor<1x10xf32>, tensor<4x10xf32>) -> tensor<4x10xf32>
  // CHECK: return %arg0
  return %sub : tensor<1x10xf32>
}
```
here we cannot replace the add by arg0 because it would change the tensor shape.

File: mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp [hunk 2]
```diff
@@ -474,7 +474,7 @@ OpFoldResult AddOp::fold(ArrayRef<Attribute> operands) {
   auto resultTy = getType().dyn_cast<RankedTensorType>();
   if (!lhsTy || !rhsTy || !resultTy)
     return {};
-  if (lhsTy != rhsTy)
+  if (lhsTy.getRank() != rhsTy.getRank())
```
Comment (TinaAMD) [inline mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp:477]: I think you don't need to check that, it would be an invalid tosa op if the [ranks don't match](https://www.mlplatform.org/tosa/tosa_spec.html#_add) (so the verifier should catch this).

**Review by mgehre-amd — APPROVED** [review]
Approved, but please handle the broadcasting case

**Review by ljfitz — APPROVED** [review]

**Review by TinaAMD — APPROVED** [review]
LGTM
